Hot Reload

jdcaperon opened the pull request:
Proposal to switch from `run watch` to hot reload and unblock updating react scripts.

https://github.com/CoEDL/elpis-gui/pull/91

Configured with `launch.json`:
```json
{
    "version": "0.2.0",
    "compounds": [
        {
            "name": "Launch w/ Webpack Dev Server",
            "configurations": ["Python: Flask", "Node: Elpis-Gui"]
        }
    ],

    "configurations": [

        {
            "name": "Python: Flask",
            "type": "python",
            "request": "launch",
            "module": "flask",
            "env": {
                "FLASK_APP": "elpis",
                "FLASK_ENV": "development",
                "FLASK_DEBUG": "0"
            },
            "args": [
                "run",
                "--host","0.0.0.0",
                "--port","5000",
                "--no-debugger",
                "--no-reload"
            ],
            "jinja": true,
        },
        {
            "name": "Node: Elpis-Gui",
            "type": "node",
            "request": "launch",
            "cwd": "/elpis-gui/",
            "env": {
                "NODE_ENV": "development",
                "NODE_PATH": "src/",
                "WDS_SOCKET_PORT": "3000"
            },
            "runtimeExecutable": "npm",
            "runtimeArgs": ["start"],
            "console": "integratedTerminal",
        }
    ]
}
```

nicklambourne commented:
Did you say this was ready for merge/review?

nicklambourne commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+ * Remote - SSH
+ * Remote - SSH: Editing Configuration Files
+ * Remote - SSH: Explorer
+ * Remote - WSL (if your working on windows)
```
you're

nicklambourne commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+
+### 5.3 Run Config
+
+Going one step further, we can setup some default run configurations so that when `F5` is pressed, the server and is debuggable from within the editor. To crate a run config to start the server, in the `.vscode` directory, create a new file called `launch.json` with the following contents:
```
`the server and is debuggable` : I think a word is missing here

nicklambourne commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+
+## 5.4 Terminal
+
+Notice that when you press ctrl-` the terminal that opens is in the container.
```
Missing backtick at the start of `ctrl-\`

nicklambourne commented on docs/development_environment.md:
````diff
@@ -0,0 +1,227 @@
+
+```json
+"forwardPorts": [
+    3000, // Forward the default webpack dev server port
````
JSON does not support comments. Don't make me revoke your plumbing certification.

nicklambourne commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+
+If there are any problems with repositories being out of sync, the best thing to do is rebuild the dev docker container from within VSCode. To do this, click the *Dev Container* option in the bottom left corner of VSCode or use the command pallet (cmd-shift-p) to find the `Remote-Containers: Rebuild container` and select it. This option will pull the newest elpis docker image.
+
+If there are further issues, check the `Dev Container` terminal (normally number 2.) (ctrl-`) for any errors that might have occurred while building the image.
```
Missing backtick at the start of `ctrl-`

nicklambourne reviewed (changes requested):
Just a few nits

benfoley commented:
will need to copy the documentation to the docs-new branch (or maybe merge that into master before merging this one?)

jdcaperon commented:
Still a little weirdness with building the container that I need to work out, let me know if it doesn't work first time on testing

benfoley reviewed (commented):
is it possible to have hot reload AND cra-build-watch or are they mutually exclusive?

nicklambourne reviewed (changes requested):
Images appear to be missing + other changes requested by Ben. Code looks fine to me.

nicklambourne commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+
+## 5. Setup Python Development Environment
+
+The `setup.sh` does all the hard work of installing linters and testing software...
```
should be `setup.py`. I had to run `python setup.py develop` to get it to install the dependencies.

nicklambourne commented on docs/development_environment.md:
````diff
@@ -0,0 +1,230 @@
+}
+```
+
+Now press `F5` (or the play button) and see the server run. Try using breakpoints to pause the server at a line of code. Try the `elpis/elpis/__init__.py` file pause the program around line 72 on the print statement in the index function. To get the server to pause you will need to open a browser and load `0.0.0.0:5000` to run that section of code.
````
<img width="698" alt="Screen Shot 2020-09-22 at 10 34 46 am" src="https://user-images.githubusercontent.com/9084563/93834465-5dbf6980-fcbf-11ea-87ef-e04baf0a742a.png">
It doesn't appear to stop at this breakpoint.

jdcaperon replied:
Ah yep, that documetnation is wrong, I'll update it, it should say browse to `0.0.0.0:5000/index.html`

jdcaperon commented on docs/development_environment.md:
```diff
@@ -0,0 +1,227 @@
+ * Remote - WSL (if your working on windows)
+
+![VSCode Extentions](assets/dev-in-vscode/vsc-extentions.png)
+
```
That should be the expected directory structure

jdcaperon commented:
> is it possible to have hot reload AND cra-build-watch or are they mutually exclusive?

I though it wasn't possible. Turns out it is, and the npm package has been updated so I'll update it on the elpis-gui repo.

p.s. when are we embracing the mono repo? 😂

nicklambourne commented:
> p.s. when are we embracing the mono repo? 😂

When your company releases its cool open source tooling that solves all of the Github scaling issues ;)

jdcaperon commented:
> When your company releases its cool open source tooling that solves all of the Github scaling issues ;)

When Elpis hits 400 concurrent developers I'm sure you'll be able to make your own 😛

jdcaperon commented:
I've updated it to use an environment variable to enable the proxy. I've updated the instructions for this. The only hiccup with my latest run-through was `requests` not being installed.

nicklambourne commented on docs/wiki/development_environment.md:
```diff
@@ -0,0 +1,308 @@
+This works by using VSCode extensions that install another layer on-top of the Elpis image that runs a VSCode server that allows the editor access to the Python runtime and seamless access to the container.
+
+![s](assets/dev-in-vscode/architecture-containers.png)
+(ref 1)
```
Can this just be `[1]`?

benfoley replied:
I've made this change.

nicklambourne reviewed (approved):
lgtm, have tested and it seems to work. I also needed to install requests manually, but that should fix itself when the next image comes out.

nicklambourne commented:
Can you merge master into this (and elpis-gui). While the hot-reloading aspect appears to work, the interface is broken and file processing appears to fail.

nicklambourne reviewed (changes requested):
See above comment

<img width="1216" alt="Screen Shot 2020-09-28 at 2 00 13 pm" src="https://user-images.githubusercontent.com/9084563/94389379-1b040280-0193-11eb-8585-71aef720d745.png">
<img width="899" alt="Screen Shot 2020-09-28 at 2 00 55 pm" src="https://user-images.githubusercontent.com/9084563/94389385-1dfef300-0193-11eb-9918-c85508596893.png">

benfoley commented:
I have run it again with the latest merge and the interface issues have been fixed. I'm happy for this to be merged.

nicklambourne commented:
I keep getting this, despite having set it up on the cinstudio macbook as per the instructions as well.

`➜  /elpis git:(jack-hottest-reload)  /usr/bin/env /venv/bin/python3 /root/.vscode-server/extensions/ms-python.python-2020.9.112786/pythonFiles/lib/python/debugpy/launcher 36311 -- -m flask run --host 0.0.0.0 --port 5000 --no-debugger --no-reload 
source /venv/bin/activate
 * Serving Flask app "elpis"
 * Environment: development
 * Debug mode: off
in index with: 
[2020-10-03 07:53:53,954] ERROR in app: Exception on / [GET]
Traceback (most recent call last):
  File "/venv/lib/python3.6/site-packages/urllib3/connection.py", line 160, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.6/http/client.py", line 1264, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1310, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1259, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1038, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 976, in send
    self.connect()
  File "/venv/lib/python3.6/site-packages/urllib3/connection.py", line 187, in connect
    conn = self._new_conn()
  File "/venv/lib/python3.6/site-packages/urllib3/connection.py", line 172, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f643ecba5c0>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 727, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/venv/lib/python3.6/site-packages/urllib3/util/retry.py", line 439, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=3000): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f643ecba5c0>: Failed to establish a new connection: [Errno 111] Connection refused',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/_compat.py", line 39, in reraise
    raise value
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/elpis/elpis/__init__.py", line 88, in index
    return proxy('http://localhost:3000/', path)
  File "/elpis/elpis/__init__.py", line 104, in proxy
    response = get(f"{host}{path}")
  File "/venv/lib/python3.6/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/venv/lib/python3.6/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/venv/lib/python3.6/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/venv/lib/python3.6/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/venv/lib/python3.6/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=3000): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f643ecba5c0>: Failed to establish a new connection: [Errno 111] Connection refused',))`

jdcaperon commented:
@nicklambourne let's debug this on zoom, you name the time

jdcaperon commented:
Oh, I know why. You need to also start the webpack dev server @nicklambourne I have a fix for this actually that I'll add to this PR. You can make a compound launch that will launch both with a single command. I still need to work out the kinks with it:
```json
"compounds": [
        {
            "name": "Launch w/ Webpack Dev Server",
            "configurations": ["Python: Flask", "Node: Elpis-Gui"]
        }
    ],
```

nicklambourne commented:
Hmm, still no joy with the compound and a rebuild. Are you free at 2pm tomorrow?

nicklambourne commented:
Have you had a chance to look at documenting those last few steps @jdcaperon?

jdcaperon commented:
Will do it today @nicklambourne

jdcaperon commented:
Done
